cmd/vsphere-xcopy-volume-populator/internal/powerstore/powerstore.go

```diff
@@ -0,0 +1,297 @@
+	klog.Infof("Volume %s is currently mapped to hosts: %v", targetLUN.Name, mappedHosts)
+	hostName := mappingContext["hostName"].(string)
+	initiatorGroup := mappingContext["initiatorGroup"].(string)
+	if slices.Contains(mappedHosts, hostName) {
+		mappedHosts = append(mappedHosts, initiatorGroup)
+	}
+	return mappedHosts, nil
+}
+
+// EnsureClonnerIgroup implements populator.StorageApi.
+func (p *PowerstoreClonner) EnsureClonnerIgroup(initiatorGroup string, adapterIds []string) (populator.MappingContext, error) {
+	klog.Infof("ensuring host %s for adapters %v", initiatorGroup, adapterIds)
+
+	ctx := context.Background()
+	mappingContext := make(map[string]any)
+
+	hosts, err := p.Client.GetHosts(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get hosts: %w", err)
+	}
+	// iterate over hosts and check if the initiator group is already mapped to the host
+	for _, host := range hosts {
+		for _, initiator := range host.Initiators {
+			for _, adapterId := range adapterIds {
+				if initiator.PortName == adapterId {
+					klog.Infof("Found existing host %s with ID %s name %s", initiatorGroup, host.ID, host.Name)
+					return populator.MappingContext{
+						"hostID":         host.ID,
+						"hostName":       host.Name,
+						"initiatorGroup": initiatorGroup,
+					}, nil
+				}
+			}
+		}
+	}
+
+	host, err := p.Client.GetHostByName(ctx, initiatorGroup)
+	if err != nil {
+		klog.Infof("Host %s not found, creating new host", initiatorGroup)
+		osType := gopowerstore.OSTypeEnumESXi
+		inits := make([]gopowerstore.InitiatorCreateModify, 0, len(adapterIds))
+		for _, a := range adapterIds {
+			pt := detectPortType(a)
+			inits = append(inits, gopowerstore.InitiatorCreateModify{
+				PortName: &a,
+				PortType: &pt,
+			})
+		}
+
+		createParams := &gopowerstore.HostCreate{
+			Name:       &initiatorGroup,
+			OsType:     &osType,
+			Initiators: &inits,
+		}
+
+		createResp, err := p.Client.CreateHost(ctx, createParams)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create host %s: %w", initiatorGroup, err)
+		}
+
+		// Get the created host
+		host, err = p.Client.GetHost(ctx, createResp.ID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get created host %s: %w", createResp.ID, err)
+		}
+
+		klog.Infof("Successfully created host %s with ID %s", initiatorGroup, host.ID)
+	} else {
+		klog.Infof("Found existing host %s with ID %s", initiatorGroup, host.ID)
+	}
+
+	// Step 2: Add initiators (adapter IDs) to the host
+	atLeastOneAdded := false
+
+	for _, adapterId := range adapterIds {
+		portType := detectPortType(adapterId)
+		klog.Infof("Processing adapter %s with type %s", adapterId, portType)
+
+		// Check if initiator already exists on this host
+		found := false
+		for _, initiator := range host.Initiators {
+			if initiator.PortName == adapterId {
+				klog.Infof("Initiator %s already exists on host %s", adapterId, initiatorGroup)
+				found = true
+				atLeastOneAdded = true
+				break
+			}
+		}
+
+		if !found {
```

Comment:
say we have a 2nd migration to a new host. we already now have a host named 'xcopy-esxs' - we treat that host object as a group effectively - we add the host  adapter to the host's initiator group field right? 
what is the behaviour for the map and unmap in that case?